# Aprendiendo del Shell

El Shell es un programa que toma comandos desde el teclado
y los pasa al Sistema Operativo para realizar alguna operación.

Casi todos las distribuciones Linux,proporcionan un shell, desde el
proyecto GNU, llamado [`bash`](https://www.gnu.org/software/bash/), que 
reemplaza al original shell de Unix, llamado `sh`.

## Emuladores de terminal

Cuando estamos usando una interfaz de usuario gráfica, necesitamos otro programa 
llamado `emulador de terminal (terminal emulador)` o simplemente  **terminal** para interactuar con el shell.

Por ejemplo para interfaces como KDE Gnome

- KDE       **konsole**.
- GNOME     **gnome-terminal.**


## Primeros pasos

Normalmente podemos lanzar el terminal escribiendo `alt + F2`

``` bash
[nombre-usuario@linux~]$
```

Eso es llamado un `prompt de Shell` y aparece siempre que el shell, está listo para
recibir entradas. Esto puede variar de la distribución y usualmente incluye
`nombre-usuario@nombre-maquina`, seguido por el actual directorio de trabajo y un signo de dolar.

Algunas veces en lugar del signo dolar, aparece el signo `#`, esto significa que el terminal tiene `previlegios de super usuario`. Eso significa que tu estas logeado como el `usuario root` o has seleccionado un terminal con permisos de `superusuario`.


Si presionas la tecla arriba (up-arrow) para recordar el comando  anterior. Esto es llamado `historia de comandos (command history)`. Presiona la tecla abajo (down-arrow) y el previo comando desaparece.

## Algunos comandos simples

- `date`: Comando que  muestra la fecha y la hora

In [1]:
date

jue ago  3 11:54:32 PET 2017


- `cal`: Muestra un calendario del mes actual.

In [2]:
cal


    Agosto 2017       
do lu ma mi ju vi sá  
       1  2  3  4  5  
 6  7  8  9 10 11 12  
13 14 15 16 17 18 19  
20 21 22 23 24 25 26  
27 28 29 30 31        
                      


- `df` Comando que sirve para ver la cantidad de espacio libre de tus particiones

In [3]:
df

S.ficheros     bloques de 1K   Usados Disponibles Uso% Montado en
udev                 1953312        0     1953312   0% /dev
tmpfs                 394740     6428      388312   2% /run
/dev/sda5           76766240 15043832    57799772  21% /
tmpfs                1973696    13556     1960140   1% /dev/shm
tmpfs                   5120        4        5116   1% /run/lock
tmpfs                1973696        0     1973696   0% /sys/fs/cgroup
/dev/sda1              98304    29519       68785  31% /boot/efi
/dev/sda8          267776736 50215704   203935688  20% /home
cgmfs                    100        0         100   0% /run/cgmanager/fs
tmpfs                 394740       60      394680   1% /run/user/1000


- `free` Muestra la cantidad de memoria libre.

Para finalizar la sesión en el terminal, cerrando la ventana del terminal o escribiendo `exit` en el prompt  del terminal.

## Navegando en el sistema de archivos


Como en Windows, Linux organiza sus archivos de manera jerarquica ` hierarchical directory structure`. Esto significa que ellos están organizados como un diseño de árboles de directorios, los cuales contienen directorios y archivos. El primer directorio en el sistema es llamado `directorio raiz(root directory)` que contiene directorios y archivos, los cuales a su vez contienen más directorios y archivos. Por ejemplo estos directorios son los que contiene un directorio raiz típico.


``` bash
bin    dev   initrd.img      lost+found  opt   run   sys  var
boot   etc   initrd.img.old  media       proc  sbin  tmp  vmlinuz
cdrom  home  lib             mnt         root  srv   usr  vmlinuz.old
```
Los sistemas Linux, siempre tienen un único sistema de archivos independiente de como sus dispositivos de almacenamiento están  unidos al computador. Los dispositivos de almacenamiento son ** montados **  a varios puntos del árbol, dependiendo de lo que se deba hacer con ellos y del administrador del sistema.

### El actual directorio de trabajo

- Para mostrar el actual `directorio de trabajo (working directory)` usamos el comando `pwd(print working directory)`

In [4]:
pwd

/home/c-lara/Escritorio


Cuando iniciamos o nos logeamos en nuestro sistema (o empezar el terminal), nuestro actual directorio de trabajo es nuestro directorio `home`. Cada cuenta de usuario tiene su propio directorio `home` y es el lugar donde normalmente se escriben los archivos.

- Para listar todos los archivos y directorios del actual directorio de trabajo, usamos el comando `ls`.

In [5]:
ls

009-slides_cache           Haroopad.desktop
009-slides_files           Introduction to R and RStudio.html
009-slides.html            libreoffice5.0-calc.desktop
009-slides.Rmd             libreoffice5.0-writer.desktop
2006.csv.short             lxterminal.desktop
Abalone.data               Notas.txt
Abalone.names              Old Firefox Data
Art of R Programming Code  OReilly.Learning.Python.5th.Edition.1449355730.pdf
ClaseR1-figure             PUMS5_06.TXT.short
ClaseR1.md                 README.txt
ClaseR1.Rpres              RNotas1.html
code.desktop               RNotas3-figure
ComandosLinux.ipynb        RNotas3.md
Datos2.ipynb               RNotas3.Rpres
Datos3.ipynb               RNotas3-rpubs.html
firefox.desktop            rstudio.desktop
FoxitReader.desktop        slides.css
g1.png                     texstudio.desktop
Grades.txt


Para cambiar el  actual directorio de trabajo, usamos el comando `cd`. Para hacer esto, escribimos `cd` seguido por la `ruta de nombres` del directorio al que queremos ir. Una `ruta de nombres` es una ruta que tomamos a lo largo de las ramas del árbol de directorios para conseguir el directorio que queremos. Las `rutas de nombres` pueden ser especificados de dos maneras distintas: rutas de nombres absolutas y relativas.

Las `rutas de nombres absoluta` inician con el directorio raiz y siguen por las ramas, hasta completar la ruta al directorio o archivo deseado.

In [6]:
cd /usr/bin

In [7]:
pwd

/usr/bin


In [8]:
ls

[                                    mtype
2to3                                 mvn
2to3-2.7                             mvnDebug
2to3-3.5                             mxtar
411toppm                             mzip
a2ping                               namei
a5toa4                               native2ascii
aa-enabled                           nawk
abiword                              ncal
aclocal                              ncurses5-config
aclocal-1.15                         ndg_httpclient
aconnect                             neotoppm
add-apt-repository                   neqn
addpart                              netkit-ftp
addr2line                            network-admin
adhocfilelist                        newgrp
afm2afm                              ngettext
afm2pl                               nice
afm2tfm                              nl
aj                                   nm
aj5                                  nm-applet
ajbrowser                            nmcli
ajc           

blueman-services                     pbmtextps
bluemoon                             pbmto10x
bluetoothctl                         pbmtoascii
bmptopnm                             pbmtoatk
bmptoppm                             pbmtobbnbg
bmtoa                                pbmtocmuwm
bootctl                              pbmtoepsi
broadwayd                            pbmtoepson
browse                               pbmtog3
brushtopbm                           pbmtogem
bsd-from                             pbmtogo
bsd-write                            pbmtoicon
bsh                                  pbmtolj
bsondump                             pbmtomacp
btattach                             pbmtomda
btmgmt                               pbmtomgr
btmon                                pbmtonokia
bugpoint-3.8                         pbmtopgm
bundledoc                            pbmtopi3
busctl                               pbmtoplot
c++                                  pbmtoppa
c2ph                  

createuser                           pgmtoppm
c_rehash                             pg_receivexlog
crontab                              pg_recvlogical
csplit                               pg_renamecluster
ctangle                              pgrep
ctanify                              pg_restore
ctanupload                           pg_upgradecluster
ctie                                 pg_virtualenv
ctstat                               pi1toppm
cups-calibrate                       pi3topbm
cupstestdsc                          pic
cupstestppd                          pico
curl                                 piconv
curl-config                          pidgin
cut                                  pidstat
cvt                                  pilconvert.py
cweave                               pildriver.py
dbus-cleanup-sockets                 pilfile.py
dbus-daemon                          pilfont.py
dbus-launch                          pilprint.py
dbus-monitor                         pinentry

dh_strip_nondeterminism              poff
dh_suidregister                      poff.wvdial
dh_testdir                           policytool
dh_testroot                          pon
dh_translations                      pon.wvdial
dh_ucf                               pooltype
dh_undocumented                      POST
dh_update_autotools_config           ppdc
dh_usrlocal                          ppdhtml
diff                                 ppdi
diff3                                ppdmerge
diffstat                             ppdpo
dig                                  pphs
dircolors                            ppltotf
dirmngr                              ppm3d
dirmngr-client                       ppmbrighten
dirname                              ppmchange
dirsplit                             ppmcie
disdvi                               ppmcolormask
display                              ppmcolors
display-im6                          ppmdim
djpeg                                ppmdist
dm-tool   

exceltex                             ptex
exifautotran                         ptftopl
expand                               ptx
expiry                               pulseaudio
explode.py                           pwdx
expr                                 py3clean
extcheck                             py3compile
extractbb                            py3versions
eyuvtoppm                            pybuild
f2py                                 pyclean
f2py2.7                              pycompile
f77                                  pydoc
f95                                  pydoc2.7
factor                               pydoc3
faillog                              pydoc3.5
faked-sysv                           pygettext
faked-tcp                            pygettext2.7
fakeroot                             pygettext3
fakeroot-sysv                        pygettext3.5
fakeroot-tcp                         pygmentex
fallocate                            pygmentize
fc-cache                         

genisoimage                          routef
genrb                                routel
geqn                                 rpcgen
GET                                  rpdfcrop
getconf                              Rscript
geteltorito                          rsh
getent                               rstart
getfacl                              rstartd
getkeycodes                          rstudio
getmapdl                             rsync
getopt                               rtstat
gettext                              ruby
gettextize                           ruby1.9.1
gettext.sh                           ruby2.1
gfortran                             ruby2.3
gfortran-4.8                         runcon
gfortran-5                           run-mailcap
gftodvi                              run-with-aspell
gftopk                               rview
gftype                               rvim
ghostscript                          sadf
gifmaker.py                          sancov-3.8
giftopnm       

gvfs-monitor-file                    ssh-keyscan
gvfs-mount                           ssindex
gvfs-move                            st4topgm
gvfs-open                            startlxde
gvfs-rename                          start-pulseaudio-x11
gvfs-rm                              startx
gvfs-save                            stat
gvfs-set-attribute                   stdbuf
gvfs-trash                           strace
gvfs-tree                            stream
h2ph                                 stream-im6
h2xs                                 strings
hardening-check                      strip
hardinfo                             sty2dtx
haroopad                             sudo
hciattach                            sudoedit
hcitool                              sudoreplay
hd                                   sum
head                                 svn-multi
HEAD                                 sylpheed
helpztags                            symcryptrun
hex2hcd                              

js                                   ttfdump
jsadebugd                            ttfread
json_pp                              tty
jstack                               typeoutfileinfo
jstat                                tzselect
jstatd                               ubuntu-bug
julia                                ubuntu-core-launcher
junit                                ubuntu-drivers
jvisualvm                            ubuntu-support-status
kbdinfo                              ucf
kbxutil                              ucfq
keytool                              ucfr
killall                              uconv
kmodsign                             ucs2any
koi8rxterm                           udisksctl
kpseaccess                           uic
kpsepath                             uic3
kpsereadlink                         ul
kpsestat                             ulqda
kpsetool                             unattended-upgrade
kpsewhere                            unattended-upgrades
kpsewhich     

lockfile-touch                       X
logger                               X11
logname                              x11perf
look                                 x11perfcomp
lorder                               x86_64
lp                                   x86_64-linux-gnu-addr2line
lpoptions                            x86_64-linux-gnu-ar
lpstat                               x86_64-linux-gnu-as
lrelease                             x86_64-linux-gnu-c++filt
lsattr                               x86_64-linux-gnu-cpp
lsb_release                          x86_64-linux-gnu-cpp-4.8
lscpu                                x86_64-linux-gnu-cpp-5
lsdiff                               x86_64-linux-gnu-dwp
lshw                                 x86_64-linux-gnu-elfedit
lsinitramfs                          x86_64-linux-gnu-g++
lsipc                                x86_64-linux-gnu-g++-4.8
lslocks                              x86_64-linux-gnu-g++-5
lslogins                             x86_64-linux-gnu-gcc
lsof

mdel                                 xhost
mdeltree                             ximtoppm
mdir                                 xinit
mdu                                  xinput
mendex                               xjc
mesg                                 xkbbell
mf                                   xkbcomp
mf2pt1                               xkbevd
mf-nowin                             xkbprint
mformat                              xkbvleds
mfplain                              xkbwatch
mft                                  xkeystone
mgrtopbm                             xkill
minfo                                xload
mkfifo                               xlogo
mkfontdir                            xlsatoms
mkfontscale                          xlsclients
mkindex                              xlsfonts
mkisofs                              xmag
mkjobtexmf                           xman
mkmanifest                           xmessage
mk_modmap                            xml2-config
mkocp           

Una ruta de nombres absoluta empieza desde el directorio raíz y recorre  la rama del árbol del sistema de archivos hasta llegar a su destino, un ruta de nombres relativa en cambio empieza desde el directorio de trabajo. Para ello, utiliza un par de símbolos para representar  posiciones relativas en el árbol del sistema de archivos. Estos símbolos especiales son `"." (punto) y ".." (punto punto)`.

Los "." símbolo se refiere al directorio de trabajo y el ".." símbolo se refiere a la directorio padre (un directorio un nivel por encima del directorio actual) del directorio de trabajo. Así es como funciona:

In [9]:
pwd

/usr/bin


Ahora queremos cambiar el directorio actual al directorio padre de `/usr/bin`, el cual es `/usr`. Podemos hacer esto de dos maneras diferentes:

In [10]:
cd /usr

In [11]:
pwd


/usr


o con una ruta de nombres relativa:




In [12]:
cd ..
pwd

/


In [13]:
cd


In [14]:
pwd


/home/c-lara


### Útiles atajos para cd

```
Atajo                        Resultado
cd                 Cambia tu directorio de trabajo a tu directorio home
cd -               Cambia tu directorio de trabajo al previo directorio de trabajo
cd ~nombre_usuario Cambia el directorio de trabajo al directorio home nombre_usuario
```

### Observaciones importantes

- Los archivos que empiezan con un punto, son archivos ocultos. Esto significa que `ls` no puede listarlos, para ello se utiliza `ls -a`, esto es debido a que  cuando se creó tu cuenta, varios archivos ocultos se colocaron en tu directorio personal para configurar  cosas por ti mismo, además de que  algunas aplicaciones colocan sus archivos de configuración y ajustes en tu directorio personal como archivos ocultos.

In [16]:
cd Escritorio
ls -a

.                          Grades.txt
..                         Haroopad.desktop
009-slides_cache           Introduction to R and RStudio.html
009-slides_files           .ipynb_checkpoints
009-slides.html            libreoffice5.0-calc.desktop
009-slides.Rmd             libreoffice5.0-writer.desktop
2006.csv.short             lxterminal.desktop
Abalone.data               Notas.txt
Abalone.names              Old Firefox Data
Art of R Programming Code  OReilly.Learning.Python.5th.Edition.1449355730.pdf
ClaseR1-figure             PUMS5_06.TXT.short
ClaseR1.md                 README.txt
ClaseR1.Rpres              RNotas1.html
code.desktop               RNotas3-figure
ComandosLinux.ipynb        RNotas3.md
Datos2.ipynb               RNotas3.Rpres
Datos3.ipynb               RNotas3-rpubs.html
firefox.desktop            rstudio.desktop
FoxitReader.desktop        slides.css
g1.png                     texstudio.desktop


- Los nombres de archivos y comandos en Linux son `case sensitive`. Los nombres de archivos `ARCHIVO1` y `archivo1` se refieren a diferentes archivos.
- Linux no tiene un concepto de una `"extensión"` como otros sistemas operativos. Los contenidos y/o función de un archivo se determina por otros medios. Aunque Linux no utiliza las extensiones de archivo para determinar el contenido/ propósito de un archivo, algunas aplicaciones lo hacen.
- Aunque Linux soporta nombres de archivo  que pueden contener espacios incrustados y caracteres de puntuación, debes limitar los caracteres de puntuación en los nombres de archivos y lo más importante, no incrustar espacios en los nombres de archivo. Si desea representar espacios entre las palabras en un nombre de archivo, utilice caracteres de subrayado.

## Explorando el sistema

`ls` es el comando más usado, y con justa razón. Con este comando podemos ver el contenido de directorios y determinar una variedad de archivos importantes y atributos de directorios.

In [17]:
ls

009-slides_cache           Haroopad.desktop
009-slides_files           Introduction to R and RStudio.html
009-slides.html            libreoffice5.0-calc.desktop
009-slides.Rmd             libreoffice5.0-writer.desktop
2006.csv.short             lxterminal.desktop
Abalone.data               Notas.txt
Abalone.names              Old Firefox Data
Art of R Programming Code  OReilly.Learning.Python.5th.Edition.1449355730.pdf
ClaseR1-figure             PUMS5_06.TXT.short
ClaseR1.md                 README.txt
ClaseR1.Rpres              RNotas1.html
code.desktop               RNotas3-figure
ComandosLinux.ipynb        RNotas3.md
Datos2.ipynb               RNotas3.Rpres
Datos3.ipynb               RNotas3-rpubs.html
firefox.desktop            rstudio.desktop
FoxitReader.desktop        slides.css
g1.png                     texstudio.desktop
Grades.txt


Podemos especificar el directorio a listar:

In [18]:
ls /boot

abi-3.13.0-125-generic         memtest86+.elf
abi-4.4.0-83-generic           memtest86+_multiboot.bin
abi-4.4.0-87-generic           System.map-3.13.0-125-generic
config-3.13.0-125-generic      System.map-4.4.0-83-generic
config-4.4.0-83-generic        System.map-4.4.0-87-generic
config-4.4.0-87-generic        vmlinuz-3.13.0-125-generic
efi                            vmlinuz-3.13.0-125-generic.efi.signed
grub                           vmlinuz-4.4.0-83-generic
initrd.img-3.13.0-125-generic  vmlinuz-4.4.0-83-generic.efi.signed
initrd.img-4.4.0-83-generic    vmlinuz-4.4.0-87-generic
initrd.img-4.4.0-87-generic    vmlinuz-4.4.0-87-generic.efi.signed
memtest86+.bin


Podemos cambiar el formato de salida para revelar más detalles.

In [20]:
ls -ls

total 12940
   4 drwxrwxr-x  3 c-lara c-lara    4096 jul 28 14:32 009-slides_cache
   4 drwxrwxr-x  3 c-lara c-lara    4096 jul 28 14:32 009-slides_files
 828 -rw-rw-r--  1 c-lara c-lara  844925 jul 28 14:32 009-slides.html
   4 -rw-rw-r--  1 c-lara c-lara    2604 jul 28 14:32 009-slides.Rmd
  32 -rw-rw-r--  1 c-lara c-lara   32467 abr 16  2012 2006.csv.short
 188 -rw-rw-r--  1 c-lara c-lara  191873 abr  2  2012 Abalone.data
   8 -rw-rw-r--  1 c-lara c-lara    4319 abr  2  2012 Abalone.names
   4 drwxr-xr-x 17 c-lara c-lara    4096 feb 16  2012 Art of R Programming Code
   4 drwxrwxr-x  2 c-lara c-lara    4096 jul 28 11:59 ClaseR1-figure
   4 -rw-rw-r--  1 c-lara c-lara     824 jul 28 14:10 ClaseR1.md
   4 -rw-rw-r--  1 c-lara c-lara     548 jul 28 14:10 ClaseR1.Rpres
   4 -rw-rw-r--  1 c-lara c-lara     101 jul 21 09:06 code.desktop
 132 -rw-rw-r--  1 c-lara c-lara  135092 ago  3 12:54 ComandosLinux.ipynb
 928 -rw-rw-r--  1 c-lara c-lara  946701 jul 26 01:20 Datos2.ipynb
 800 -rw-rw-r

### Opciones y argumentos

Los  comandos suelen ir acompañados de una o más opciones que modifican su comportamiento y además, por uno o más argumentos, sobre los que los comandos actuan. Así  la mayoría de los comandos se ven más o menos así:

```
comandos -opciones argumentos
```

La mayoría de los comandos utilizan opciones consisten  de un sólo carácter precedido por un guión, por ejemplo, `"-l"`, pero muchos comandos, incluidos los del proyecto GNU, también admiten opciones largas, que consisten en una palabra precedida de dos guiones. Además, muchos comandos permiten múltiples opciones cortas que se encadenan juntas.

En este ejemplo, el comando `ls` tiene dos opciones, la opción "l" para producir la salida de formato largo, y la opción "t" para ordenar el resultado por fecha de modificación del archivo.

In [21]:
ls -lt

total 12944
-rw-rw-r--  1 c-lara c-lara  138689 ago  3 12:54 ComandosLinux.ipynb
-rw-rw-r--  1 c-lara c-lara  163042 ago  2 11:27 RNotas3-rpubs.html
-rw-rw-r--  1 c-lara c-lara   13280 ago  2 11:22 RNotas3.md
-rw-rw-r--  1 c-lara c-lara   13019 ago  2 11:22 RNotas3.Rpres
-rw-rw-r--  1 c-lara c-lara   30293 ago  2 09:31 g1.png
drwxrwxr-x  2 c-lara c-lara    4096 ago  2 09:13 RNotas3-figure
-rw-rw-r--  1 c-lara c-lara  908856 jul 30 22:46 RNotas1.html
-rw-rw-r--  1 c-lara c-lara  815109 jul 29 13:32 Datos3.ipynb
-rw-rw-r--  1 c-lara c-lara     397 jul 29 00:34 Notas.txt
-rw-rw-r--  1 c-lara c-lara 1116795 jul 28 15:12 Introduction to R and RStudio.html
-rw-rw-r--  1 c-lara c-lara  844925 jul 28 14:32 009-slides.html
drwxrwxr-x  3 c-lara c-lara    4096 jul 28 14:32 009-slides_files
drwxrwxr-x  3 c-lara c-lara    4096 jul 28 14:32 009-slides_cache
-rw-rw-r--  1 c-lara c-lara    2604 jul 28 14:32 009-slides.Rmd
-rw-rw-r--  1 c-lara c-lara     824 jul 28 14:10 ClaseR1.md
-rw-rw-r--  1 c-lara

Podemos agregar una opción larga `--reverse` para revertir el orden de búsqueda.

In [22]:
ls -lt --reverse

total 12944
drwxr-xr-x 17 c-lara c-lara    4096 feb 16  2012 Art of R Programming Code
-rw-rw-r--  1 c-lara c-lara  191873 abr  2  2012 Abalone.data
-rw-rw-r--  1 c-lara c-lara    4319 abr  2  2012 Abalone.names
-rw-rw-r--  1 c-lara c-lara     273 abr  2  2012 Grades.txt
-rw-rw-r--  1 c-lara c-lara   32467 abr 16  2012 2006.csv.short
-rw-rw-r--  1 c-lara c-lara   28665 abr 16  2012 PUMS5_06.TXT.short
-rw-rw-r--  1 c-lara c-lara     688 abr 16  2012 README.txt
-rw-rw-r--  1 c-lara c-lara 7827077 ago  6  2013 OReilly.Learning.Python.5th.Edition.1449355730.pdf
-rw-------  1 c-lara c-lara     455 oct 30  2015 rstudio.desktop
-rw-------  1 c-lara c-lara    9177 oct 30  2015 firefox.desktop
-rw-------  1 c-lara c-lara   27144 oct 31  2015 libreoffice5.0-writer.desktop
-rw-------  1 c-lara c-lara   22124 oct 31  2015 libreoffice5.0-calc.desktop
-rw-------  1 c-lara c-lara     384 nov  1  2015 texstudio.desktop
-rw-------  1 c-lara c-lara     226 nov  5  2015 Haroopad.desktop
drwx------  3 c-l

### Viendo la salida de formato largo

Expliquemos que es cada campo en la salida de uno de los archivos, dado por `ls -l`:

```bash
-rw-rw-r--   1 c-lara c-lara     397 jul 29 00:34 Notas.txt
-drwx------  3 c-lara c-lara    4096 jul 19 14:31 Old Firefox Data
```

- `-rw-rw-r--` o `drwx------`: Derechos de acceso al archivo. El primer carácter indica el tipo de archivo. Entre los diferentes tipos, un guión  significa un archivo normal, mientras que una "d" indica un directorio.
Los siguientes tres caracteres son los derechos de acceso para el propietario del archivo, los siguientes tres son para los miembros del grupo con acceso al archivo, y la final tres son los permisos para el resto.

- `1 o 2`: Número de `enlaces fuertes (hard links).`
- `c-lara` : El usuario propietario del archivo.
- `c-lara` : El nombre del grupo que posee el archivo.
- `397 o 4096`    : Tamaño del archivo en bytes.
- ` jul 29 00:34 jul 19 14:31 `: Fecha y hora de la última modificación del archivo.
- `Notas.txt, Old Firefox Data`: Nombre de los archivos.


### Otros comandos importantes

- `file` Comando para determinar el tipo de archivo.

In [23]:
file Notas.txt

Notas.txt: ASCII text


- `less` Este comando es un programa para ver archivos de texto y es muy útil  debido a que muchos de los archivos (**scripts **) que contienen las configuraciones del sistema (llamados archivos de configuración) se almacenan en este formato, y ser capaz de leerlos nos da una idea de cómo funciona el sistema.

```
less app.R
```

### ¿ Qué se entiende por texto?

Hay muchas formas de representar la información en un ordenador. Todos los métodos implican la definición de una relación entre la información y algunos números que se utilizarán para representarla. Las computadoras después de todo, sólo entienden números y todos los datos son convertidos  en representación numérica.

Algunos de estos sistemas de representación son muy complejos  mientras que otros son muy  simples. Uno de los primeros y más simples es llamado  `ASCII` que es la abreviatura de `American Standard Code for Information Interchange`. Este es un esquema de codificación simple que fue utilizado  para mapear los caracteres del teclado a números. El texto es un simple mapeo uno a uno de caracteres a  números. Es muy compacto. Cincuenta caracteres de texto se traducen en  cincuenta bytes de datos. Es importante entender que sólo el texto contiene un mapeo de caracteres a números. No es lo mismo que un documento de procesador de textos, como el creado por OpenOffice.org o Microsoft Word. Esos archivos, en contraste con ASCII  contienen muchos elementos no son de texto que se utilizan para describir su estructura y formato. Archivos de texto plano ASCII contienen  algunos códigos de control adicionales   como saltos de línea, tabulador, etc. A través de un sistema Linux, muchos archivos se almacenan en formato de texto y hay muchas herramientas de Linux que trabajan con archivos de texto.

Es importante conocer los archivos de texto, por que muchos archivos llamados de `configuración` son almacenados en este formato y siendo capaz de leer estos nos da una visión de como el sistema trabaja, además de que muchos programas que el sistema  usa llamados `scripts` utilizan este formato.

El comando `less` es usada como es esto

```
less nombre-archivo
```

Una vez iniciado, el programa `less` le permite desplazarse hacia adelante y hacia atrás mediante un archivo de texto. Como el siguiente ejemplo:

```
less /etc/vim/vimrc
```

La siguiente tabla muestra los comandos de teclado más comunes utilizadas por `less`


```
     Comando					Acción
   Page Up or b            Desplazarse hacia atrás una página
   Page Down or space      Desplazarse a la página siguiente
   Up Arrow                Desplazarse una línea hacia arriba
   Down Arrow              Desplazarse una línea hacia abajo
      G                    Desplazarse al final del archivo de texto
   1G or g               Ir al principio del archivo de texto
 /caracter            Busca la próxima ocurrencia de caracter
     n                  Busca la siguiente aparición de la búsqueda anterior
     h                      Visualiza la pantalla de ayuda
     q                         Quitamos less
```

El programa `less` fue diseñado como una mejora de un programa anterior llamada `more`. El nombre `"less"` es un juego en la frase `"less is more"`.

`less` cae en la clase de programas llamados `pagers` que permiten la fácil visualización de documentos de texto grandes  página por página hacia adelante y hacia atrás junto con otras características, como indica el cuadro anterior.

Podemos ver la documentación de  `less`  escribiendo `man less`.

## Directorios Importantes

El diseño del sistema de archivos en los sistemas Linux es muy similar al de  otros sistemas Unix. El diseño es en realidad especificado en un estándar llamado  `Linux Filesystem Hierarchy Standard `y mucho de los más interesantes archivos están  en archivo de texto.

Veamos una lista de los directorios más importantes del sistema de archivos de Linux

- `/`: Directorio raiz. Donde todo empieza.
- `/bin`:  Contiene programas (binarios) que deben estar presente en el sistema para que arranque y corra.
- `/boot`: Contiene el Kernel de Linux, disco de imagen RAM inicial y el gestor de arranque. Un archivo importante 	es `/boot/vmlinuz` que es el Kernel Linux.
- `/dev`: Este es un directorio especial que contiene nodos de dispositivos. "Todo es un archivo" también se aplica a los dispositivos. Aquí es donde el kernel mantiene una lista de todos los dispositivos que entiende.
- `/etc`: Es un directorio contiene todos los archivos de configuración de todo el sistema. También contiene una colección de scripts del shell que  inician cada uno de los servicios del sistema en el arranque. Todo en este directorio debe ser  texto legible. Archivos importantes:
  - `/etc/crontab`: Un archivo que define cuando un proceso automático deberá ejecutarse.
  - `/etc/fstab`:  Es una tabla de dispositivos de almacenamiento y sus puntos de montajes asociados.
  - `/etc/passwd`: Contiene una lista  de las cuentas de usuario.
- `/home`: Directorio del usuario.
- `/lib`: Contiene librerias compartidas usadas por los programas centrales del sistema. Son similares a las DLL en windows.
- `/lost + found`: Cada partición formateada o dispositivo usando el sistema de archivos  **ext3** tiene este directorio. Es usado en el caso de una recuperación parcial de un evento de sistemas dañados.
- `/media`: Es el directorio, que contiene los puntos de montaje de dispositivos removibles como los USB, CD-ROMs, etc, que son montados automáticamante en la insercción.
- `/mnt`: Contiene los puntos de montaje de dispositivos removibles que han sidos montados manualmente.
- `/opt`: Es un directorio usado para instalar *opcional* software.
- `/proc`: Es un sistema de archivo virtual mantenido por el Kernel. Los "archivos" que contiene son rendijas en el propio Kernel. Los archivos se pueden leer y le dará una idea de cómo el Kernel ve su computadora.
- `/root`: Este es el directorio de la cuenta del usuario `root`.
- `/sbin`: Contiene *sistemas binarios*, que son programas que llevan tareas vitales que son reservadas al usuario root.
- `/tmp`: El directorio `/tmp` está destinado para el almacenamiento de archivos temporales, creados por diversos programas. Algunas configuraciones causan que este directorio deba ser vaciado cada vez que se reinicia el sistema.
- `/usr`: El árbol de directorios `/usr` probablemente es el más grande en un sistema Linux. Contiene todos los archivos de los programas y de soporte utilizados por los usuarios regulares.

  - `/usr/bin`: Contiene ejecutables instalados en tu distribución.
  - `/usr/lib`: Librerias compartidas por los programas en `/usr/bin`
  - `/usr/local`: Es el directorio donde se encuentran los programas que no son incluidos con la distribución, pero que están destinados para el uso de todo el sistema. Programas compilados desde el código fuente normalmente se instalan en `/usr/local/bin`.
  - `/usr/sbin`: Contiene muchos programas de administración del sistema.
  - `/usr/share`: Es un directorio que contiene toda los datos, usados por los programas en `/usr/bin`. Esto incluye cosas como archivos de configuración, íconos, archivos de sonido, fondos de pantalla, etc.
  - `/usr/share/doc`: Documentación de programas del sistema.
- `/var`: El árbol de directorios `/var` es donde se almacenan los datos que  probablemente  cambien. Varias bases de datos,  correo de usuario, etc., se encuentran aquí.
  - `/var/log`: Contiene los archivos de registro, registros de diversas actividades del sistema. Estos son muy importantes y deben ser vigilados de vez en cuando. El archivo más útil es `/var/log/messages`. Ten en cuenta que por razones de seguridad en algunos sistemas, debe ser superusuario quien pueda ver los archivos de registro.



##  Enlaces símbolicos

Al mirar en nuestro sistema de archivos, es probable que veamos un directorio con una entrada como esta:

```
lrwxrwxrwx  1 root root  19 oct 15  2013 libsidplay.so.1 -> libsidplay.so.1.0.3

```

Observa cómo la primera letra de la lista es "l" y la entrada parece tener dos nombres de archivo? Este es un tipo especial de un archivo llamado un `enlace simbólico`.  En sistemas Unix  es posible tener un archivo referenciado por varios nombres. Mientras que el valor de este puede no ser obvio, es realmente una característica útil.

Imagina este escenario: Un programa requiere el uso de un recurso compartido de algún tipo contenido en un archivo llamado `"Mila"`, pero `"Mila"` tiene cambios de versión frecuentes. Sería bueno incluir el número de versión en el nombre del archivo para que el administrador u otra parte interesada pueda ver qué versión de `"Mila"` está instalada. Esto presenta un problema. Si cambiamos el nombre del recurso compartido, tenemos que localizar  cada programa que lo utiliza y modificarlo para buscar un nuevo nombre de recurso cada vez que se instala una nueva versión. Eso no suena divertido en absoluto.

Aquí es donde los enlaces simbólicos interviene. Digamos que instalamos la versión 2.6 de `"Mila"`, que tiene el nombre de archivo `"Mila-2.6"` y luego creamos un enlace simbólico llamado simplemente `"Mila"` que apunta a `"Mila-2.6"`. Esto significa que cuando un programa se abre el archivo `"Mila"`, en realidad está abriendo el archivo `"Mila-2.6"`. Ahora todo el mundo es feliz. Los programas que dependen de `"foo"` pueden encontrarlo y podemos ver qué versión actual está instalada. Cuando es el momento de actualizar a `"Mila-2.7"` solo  agregamos el archivo a nuestro sistema, eliminamos  el vínculo simbólico `"Mila"` y crear uno nuevo que apunta a la nueva versión. Esto no sólo resuelve el problema de la actualización de la versiones, sino que también nos permite mantener ambas versiones en nuestra máquina. Imagínese que `"Mila-2.7"` tiene un error,  tenemos que volver a la versión antigua necesariamente.

Una vez más, solo  eliminamos el vínculo simbólico que apunta a la nueva versión y creamos un nuevo enlace simbólico que apunta a la versión anterior.

## Manipulando Archivos y Directorios

### Wildcards (comodines)

Desde que el shell utiliza nombres de archivo demasiadas veces, proporciona caracteres especiales para ayudar a especificar rápidamente grupos de nombres de archivos. Estos caracteres especiales son llamados [comodines](http://www.linfo.org/wildcard.html).

Los comodines nos permiten seleccionar los nombres de archivo basado en patrones de caracteres. La siguiente tabla muestra una lista de comodines y que es lo que seleccionan

```
Comodin              Significado
*                    Enlaza uno o más caracteres
?                    Enlaza un único caracter
[caracteres]         Enlaza algún caracter que es miembro de *caracteres*
[!caracter]          Enlaza algún caracter que no es miembro de *caracteres*
[[:clase:]]         Enlaza algún caracter que es miembro de una determinada clase

```

Las clases comunmente usadas son

```
Clase de caracteres     Significado
[:alnum:]          Enlaza algún caracter alfa numérico
[:alpha:]          Enlaza algúb caracter alfabético
[:digit:]          Enlaza algún numeral
[:lower:]          Enlaza alguna letra minúscula
[:upper:]          Enlaza alguna letra mayúscula
```

## Ejemplos
- Muestra la información de cada objeto en el directorio actual.

In [10]:
file *

C:      directory
Java:   directory
Python: directory
R:      directory


- Nos movemos a un directorio y le decimos al comando `ls` que proporciones todos los nombres de los archivos en el actual directorio que tiene una extensión `.md`, `.R`, `.py`


In [13]:
pwd

/home/c-lara/codigos


In [14]:
cd ..
cd Escritorio
ls  *.md *.R *.py

app.R  ClaseR1.md  Ejemplo.py  Ejemplo.R  RNotas3.md


- Retorna una lista de archivos en el actual directorio de trabajo que tienen dos caracteres los archivos de extensión.

In [26]:
ls *.??

ClaseR1.md  Ejemplo.py  RNotas3.md


### Trabajando con Comandos

Los comandos pueden ser cualquiera de estos tipos

- `Un programa ejecutable` como los de `/usr/bin`. Dentro de esta categoria, los programas pueden ser `compilados binarios` tales como los programas en C y C++ o programas escritos en `lenguajes interpretados`` como perl, Python o Ruby.
-  `Un comando dentro del shell`. `bash` soporta un número de comandos internamente llamados `shell builtins`. El comando `cd` por ejemplo es un `shell builtins`.
-  Una `función del Shell` que son scripts de shell incorporados al `entorno (environment)`.
-  Un `alias` que son comandos que podemos definir, desde otros comandos propios del shell de Linux.

## Identificando comandos

### El comando type

El comando `type` muestra el tipo de comando que el shell ejecuta, dando un particular nombre.

```
type comando
```

donde `comando` es el nombre del comando que se quieres analizar.

In [27]:
type ls

ls es un alias de `ls --color=auto'


In [28]:
type top

top is /usr/bin/top


In [29]:
type cd 

cd es una orden interna del shell


### El comando which

`which` es un comando que muestra la localización de los archivos ejecutables, si se intenta otro tipo de comando se genera un error.

In [31]:
which cd

: 1

###  Creando directorios

El comando `mkdir` es usado para crear directorios. Y se usa de la siguiente manera, (los tres puntos, significan que el argumento puede ser repetido)

```
mkdir directorio...
```

Asi podemos crear un solo directorio o varios directorios

```
mkdir dir1
```

o

```
mkdir dir1 dir2 dir3
```

### Copiando archivos y directorios

El comando `cp` copia archivos o directorios. Puede ser usado de dos formas

```
cp item1 item2
```

para copiar el archivo  o directorio `item1` al archivo o directorio `item2` y

```
cp item ...directorio
```

para copiar múltiples ítems (archivo o directorios) en un directorio.

### Algunas opciones y ejemplos

Aquí algunas opciones para `cp`:

```
    Opciones                       Significado
-a, --archive         Copia los archivos y directorios y todo sus atributos.
-i, --interactive     Antes de sobreescribir un archivo, se pide una confirmacion.
-r, --recursive       Recursivamente copia directorios y sus contenidos.
-u, --update          Solamente copia los archivos que, o bien no existen, o son más recientes.
-v, --verbose         Muestra mensajes informativos, como que la copia se ha llevado a cabo.

```

## Ejemplo

### Creando Directorios

In [1]:
cd
mkdir codigos
cd codigos
mkdir Python R  C Java

In [2]:
pwd

/home/c-lara/codigos


In [3]:
cp /home/c-lara/Documentos/Ejemplo.R .
cp /home/c-lara/Documentos/Ejemplo.Java .
cp /home/c-lara/Documentos/Ejemplo.py .

In [4]:
ls -l

total 28
drwxrwxr-x 2 c-lara c-lara 4096 ago  3 19:26 C
-rw-rw-r-- 1 c-lara c-lara  119 ago  3 19:35 Ejemplo.Java
-rw-rw-r-- 1 c-lara c-lara  338 ago  3 19:35 Ejemplo.py
-rw-rw-r-- 1 c-lara c-lara 1322 ago  3 19:35 Ejemplo.R
drwxrwxr-x 2 c-lara c-lara 4096 ago  3 19:26 Java
drwxrwxr-x 2 c-lara c-lara 4096 ago  3 19:26 Python
drwxrwxr-x 2 c-lara c-lara 4096 ago  3 19:26 R


### Moviendo y Renombrando archivos


In [5]:
mv Ejemplo.py Python
mv Ejemplo.Java Java
mv Ejemplo.R R

### Documentación de Comandos

###  Consiguiendo ayuda para órdenes internas

Bash tiene un `help` para cada `orden interna`. Para usar esto escribimos `help` seguido por el nombre de la orden interna del shell. Muchos programas ejecutables soportan la opción `--help` que muestra la descripción del comando y opciones.

In [6]:
grep --help

Uso: grep [OPCIÓN]... PATRÓN [ARCHIVO]...
Busca PATRÓN en cada FICHERO o en la entrada estándar.
PATRÓN es, por omisión, una expresión regular básica (BRE).
Ejemplo: grep -i 'hello world' menu.h main.c

Selección e interpretación de Expreg:
  -E, --extended-regexp     PATTERN is an extended regular expression (ERE)
  -F, --fixed-strings       PATTERN is a set of newline-separated strings
  -G, --basic-regexp        PATTERN is a basic regular expression (BRE)
  -P, --perl-regexp         PATTERN is a Perl regular expression
  -e, --regexp=PATRÓN       utiliza PATRÓN como expresión regular
  -f, --file=FICHERO        obtiene PATRÓN de FICHERO
  -i, --ignore-case         considera iguales mayúsculas y minúsculas
  -w, --word-regexp         obliga a que PATRÓN coincida solamente
                            con palabras completas
  -x, --line-regexp         obliga a que PATRÓN coincida solamente
                            con líneas completas
  -z, --null-data           una línea de datos t

### El comando man

Muchos programas  ejecutables usan línea de comandos para mostrar  documentación, llamada `manual` o `man page`. Para este fin, se usa un programa llamado `man` y que se puede utilizar de la siguiente manera:

In [7]:
man less

LESS(1)                     General Commands Manual                    LESS(1)

NAME
       less - opposite of more

SYNOPSIS
       less -?
       less --help
       less -V
       less --version
       less [-[+]aABcCdeEfFgGiIJKLmMnNqQrRsSuUVwWX~]
            [-b space] [-h lines] [-j line] [-k keyfile]
            [-{oO} logfile] [-p pattern] [-P prompt] [-t tag]
            [-T tagsfile] [-x tab,...] [-y lines] [-[z] lines]
            [-# shift] [+[+]cmd] [--] [filename]...
       (See  the  OPTIONS section for alternate option syntax with long option
       names.)

DESCRIPTION
       Less is a program similar to more (1), but it has many  more  features.
       Less  does  not  have to read the entire input file before starting, so
       with large input files it starts up faster than text  editors  like  vi
       (1).  Less uses termcap (or terminfo on some systems), so it can run on
       a variety of terminals.  There is even  limited  support  for  hardcopy
       termina


       ^X^X   Same as single quote.

       /pattern
              Search forward in the file for the N-th line containing the pat‐
              tern.  N defaults to 1.  The pattern is a regular expression, as
              recognized by the regular expression library  supplied  by  your
              system.   The search starts at the first line displayed (but see
              the -a and -j options, which change this).

              Certain characters are special if entered at  the  beginning  of
              the  pattern;  they modify the type of search rather than become
              part of the pattern:

              ^N or !
                     Search for lines which do NOT match the pattern.

              ^E or *
                     Search multiple files.  That is, if  the  search  reaches
                     the  END of the current file without finding a match, the
                     search continues in the next file  in  the  command  line
                     list.

              ting is printed and nothing is changed.

       --     Like  the  -  command, but takes a long option name (see OPTIONS
              below) rather than a single option letter.  You must press ENTER
              or  RETURN after typing the option name.  A ^P immediately after
              the second dash suppresses printing of a message describing  the
              new setting, as in the - command.

       -+     Followed  by  one  of  the command line option letters this will
              reset the option to its default  setting  and  print  a  message
              describing  the  new  setting.  (The "-+X" command does the same
              thing as "-+X" on the command line.)  This  does  not  work  for
              string-valued options.

       --+    Like  the -+ command, but takes a long option name rather than a
              single option letter.

       -!     Followed by one of the command line option  letters,  this  will
              reset  the  optio

              automatically as needed.  If a large amount of data is read from
              the  pipe,  this  can cause a large amount of memory to be allo‐
              cated.  The -B option disables this automatic allocation of buf‐
              fers for pipes, so that only 64 K (or the amount of space speci‐
              fied by the -b option) is used for the pipe.  Warning: use of -B
              can  result  in  erroneous display, since only the most recently
              viewed part of the piped data is kept  in  memory;  any  earlier
              data is lost.

       -c or --clear-screen
              Causes  full  screen  repaints  to  be painted from the top line
              down.  By default, full screen repaints are  done  by  scrolling
              from the bottom of the screen.

       -C or --CLEAR-SCREEN
              Same as -c, for compatibility with older versions of less.

       -d or --dumb
              The -d option suppresses the error message normall

              below).

       -N or --LINE-NUMBERS
              Causes a line number to be displayed at the  beginning  of  each
              line in the display.

       -ofilename or --log-file=filename
              Causes  less  to copy its input to the named file as it is being
              viewed.  This applies only when the input file is a pipe, not an
              ordinary  file.   If  the file already exists, less will ask for
              confirmation before overwriting it.

       -Ofilename or --LOG-FILE=filename
              The -O option is like -o, but it will overwrite an existing file
              without asking for confirmation.

              If  no log file has been specified, the -o and -O options can be
              used from within less to specify a log  file.   Without  a  file
              name, they will simply report the name of the log file.  The "s"
              command is equivalent to specifying -o from within less.

       -ppattern or --patter


       -xn,... or --tabs=n,...
              Sets  tab  stops.  If only one n is specified, tab stops are set
              at multiples of n.  If multiple values separated by  commas  are
              specified,  tab  stops are set at those positions, and then con‐
              tinue with the same spacing  as  the  last  two.   For  example,
              -x9,17  will  set  tabs  at  positions  9, 17, 25, 33, etc.  The
              default for n is 8.

       -X or --no-init
              Disables sending the termcap initialization and deinitialization
              strings  to  the  terminal.   This is sometimes desirable if the
              deinitialization string does something unnecessary, like  clear‐
              ing the screen.

       -yn or --max-forw-scroll=n
              Specifies a maximum number of lines to scroll forward.  If it is
              necessary to scroll forward more than n  lines,  the  screen  is
              repainted  instead.   The -c or -C option

              mand which begins with that text.

       DOWNARROW [ ESC-j ]
              Retrieve  the  next  command line.  If you first enter some text
              and then press DOWNARROW, it  will  retrieve  the  next  command
              which begins with that text.

       TAB    Complete  the partial filename to the left of the cursor.  If it
              matches more than one filename, the first match is entered  into
              the  command  line.   Repeated  TABs  will  cycle thru the other
              matching filenames.  If the completed filename is a directory, a
              "/"  is  appended to the filename.  (On MS-DOS systems, a "\" is
              appended.)  The environment variable LESSSEPARATOR can  be  used
              to specify a different character to append to a directory name.

       BACKTAB [ ESC-TAB ]
              Like, TAB, but cycles in the reverse direction thru the matching
              filenames.

       ^L     Complete the partial fi

       interpreted as meaning there is no replacement, and the  original  file
       is used.  To avoid this, if LESSOPEN starts with two vertical bars, the
       exit status of the script becomes meaningful.  If the  exit  status  is
       zero,  the  output  is  considered  to  be replacement text, even if it
       empty.  If the exit status is nonzero, any output is  ignored  and  the
       original  file  is  used.   For compatibility with previous versions of
       less, if LESSOPEN starts with only one vertical bar, the exit status of
       the preprocessor is ignored.

       When  an input pipe is used, a LESSCLOSE postprocessor can be used, but
       it is usually not necessary since there is no replacement file to clean
       up.   In  this  case, the replacement file name passed to the LESSCLOSE
       postprocessor is "-".

       For compatibility with previous versions of less, the input  preproces‐
       sor or pipe is not used if less is viewing standard input

       nary  user need not understand the details of constructing personalized
       prompt strings.

       A percent sign followed by a single character is expanded according  to
       what the following character is:

       %bX    Replaced  by the byte offset into the current input file.  The b
              is followed by a single character (shown as X above) which spec‐
              ifies  the line whose byte offset is to be used.  If the charac‐
              ter is a "t", the byte offset of the top line in the display  is
              used, an "m" means use the middle line, a "b" means use the bot‐
              tom line, a "B" means use the line just after the  bottom  line,
              and  a  "j"  means use the "target" line, as specified by the -j
              option.

       %B     Replaced by the size of the current input file.

       %c     Replaced by the column number of the text appearing in the first
              column of the screen.

       %dX    Replaced

       is  invoked via a file link named "more", less behaves (mostly) in con‐
       formance with the POSIX "more" command specification.   In  this  mode,
       less behaves differently in these ways:

       The  -e  option  works  differently.  If the -e option is not set, less
       behaves as if the -e option were set.  If the -e option  is  set,  less
       behaves as if the -E option were set.

       The  -m  option  works  differently.   If the -m option is not set, the
       medium prompt is used, and it is prefixed with the  string  "--More--".
       If the -m option is set, the short prompt is used.

       The  -n  option acts like the -z option.  The normal behavior of the -n
       option is unavailable in this mode.

       The parameter to the -p option is taken to be  a  less  command  rather
       than a search pattern.

       The  LESS  environment  variable  is  ignored, and the MORE environment
       variable is used in its place.

ENVIRONMENT VARIABLES


El manual que `man` muestra es dividido en secciones y no solo cubre comandos de usuarios sino también comandos de sistemas de administración, archivos de formato, etc. La organización de la página mostrada por `man` es

```
Sección               Contenido
   1             Comandos de usuario
   2             Llamadas de interfaces de programación del kernel
   3             Programación de interfaces para la biblioteca C
   4             Archivos especiales tales como nodos y controladores de  dispositivos
   5             Formatos de archivos
   6             Juegos y entretenimientos tales como protectores de pantalla
   7             Miscélanea
   8             Comandos de administración de sistema
```

Un ejemplo de esto es el siguiente que muestra los formatos del archivo `/etc/passwd`.


In [8]:
man 5 passwd

PASSWD(5)                File Formats and Conversions                PASSWD(5)

NAME
       passwd - the password file

DESCRIPTION
       /etc/passwd contains one line for each user account, with seven fields
       delimited by colons (“:”). These fields are:

       ·   login name

       ·   optional encrypted password

       ·   numerical user ID

       ·   numerical group ID

       ·   user name or comment field

       ·   user home directory

       ·   optional user command interpreter

       The encrypted password field may be blank, in which case no password is
       required to authenticate as the specified login name. However, some
       applications which read the /etc/passwd file may decide not to permit
       any access at all if the password field is blank. If the password field
       is a lower-case “x”, then the encrypted password is actually stored in
       the shadow(5) file instead; there must be a corresponding line in the
       /etc/shadow file, or els

### El comando apropos

También es posible buscar en la lista de páginas de `man` para posibles coincidencias de un término que estamos buscando.He aquí un ejemplo de una búsqueda en las páginas de `man` utilizando el término de búsqueda `gcc`.

In [9]:
apropos gcc

c89-gcc (1)          - ANSI (1989) C compiler
c99-gcc (1)          - ANSI (1999) C compiler
gcc (1)              - GNU project C and C++ compiler
gcc-4.8 (1)          - GNU project C and C++ compiler
gcc-5 (1)            - GNU project C and C++ compiler
gcc-ar (1)           - a wrapper around ar adding the - plugin option
gcc-ar-4.8 (1)       - a wrapper around ar adding the - plugin option
gcc-ar-5 (1)         - a wrapper around ar adding the - plugin option
gcc-nm (1)           - a wrapper around nm adding the - plugin option
gcc-nm-4.8 (1)       - a wrapper around nm adding the - plugin option
gcc-nm-5 (1)         - a wrapper around nm adding the - plugin option
gcc-ranlib (1)       - a wrapper around ranlib adding the - plugin option
gcc-ranlib-4.8 (1)   - a wrapper around ranlib adding the - plugin option
gcc-ranlib-5 (1)     - a wrapper around ranlib adding the - plugin option
x86_64-linux-gnu-gcc (1) - GNU project C and C++ compiler
x86_64-linux-gnu-gcc-4.8 (1) - GNU project C a

El  primer campo en la línea de salida es el nombre de la página de `man`, el segundo campo muestra la sección. `man` con la opción `k` lleva a cabo la misma función que `apropos`

###  El comando whatis

`whatis` muestra el nombre y una línea de descripción de la página de  `man` del término buscado.


In [32]:
whatis make

make (1)             - GNU make utility to maintain groups of programs


### README y otros archivos de documentación de programas

Muchos paquetes de software instalados en el sistema tienen archivos de documentación que residen en el directorio `/usr/share/doc`. La mayoría de éstos se almacenan en formato de texto plano y puede ser visto con `less`. Algunos de los archivos están en formato HTML y se puede ver con un navegador web. 
Podemos encontrar algunos archivos que terminan con la extensión `".gz"`, esto indica que se han comprimido con el programa de compresión `gzip`. El paquete gzip incluye una versión especial de `less` llamado `zless` que muestra el contenido del archivo comprimido `gzip`.
